src/herder/Upgrades.h

```diff
@@ -75,7 +76,7 @@ class Upgrades
 
     // create upgrades for given ledger
     std::vector<LedgerUpgrade> createUpgradesFor(LedgerHeader const& lclHeader,
-                                                 AbstractLedgerTxn& ltx) const;
+                                                 LedgerStateGetter& ls) const;
```

Comment:
thanks, all good points. I kept function signatures in the Upgrades class as-is, just swapped LedgerTxn to LedgerStateGetter in some places. I agree it's weird to pass ledger header separately, so I think there's an opportunity for cleanup. Ok if we do it in a follow-up? I can open an issue.

> Just please be careful because some of these update ledger header, so they may legitimately want LCL header instead of the current header. Not this particular method though.

Yeah, we need to be very careful here. I believe I kept LedgerTxn in all the functions that actually modify the ledger. The only functions that got modified use read-only ledger state (so they either didn't take LedgerTxn at all, or used loadWithoutRecord)